c++/src/kj/memory.h

```diff
@@ -656,6 +657,275 @@ class SpaceFor {
+
+private:
+  KJ_DISALLOW_COPY_AND_MOVE(Pin);
+  
+  kj::Own<T> t;
+#ifdef KJ_ASSERT_PIN_PTRS  
+  uint64_t ptrs = 0;
+#endif
+
+  inline T* get() { return t.get(); }
+  inline const T* get() const { return t.get(); }
+
+  template <typename X, typename... Params>
+  friend Pin<kj::Own<X>> heapPin(Params&&... params);
+
+  template <typename X>
+  friend class Ptr;
+};
+
+template <typename T, typename... Params>
+inline Pin<kj::Own<T>> heapPin(Params&&... params) {
+  return Pin<kj::Own<T>>(kj::heap<T>(kj::fwd<Params>(params)...));
+}
+
+// =======================================================================================
+// Uniq<T>
+
+template <typename T>
+class Uniq {
+  // Uniq<T> is a weaker version of Pin<T>: you can move it as long as
+  // there are no live Ptr<T> pointing at it.
+
+public:
+  inline Uniq(Uniq<T>&& other): t(kj::mv(other.t)) {
+#ifdef KJ_ASSERT_PIN_PTRS  
+    KJ_IREQUIRE(other.ptrs == 0, "Uniq with active references moved");
+#endif
+  }
+
+  inline ~Uniq() {
+#ifdef KJ_ASSERT_PIN_PTRS  
+    KJ_IREQUIRE(ptrs == 0, "Uniq with active references destroyed");
+#endif
+  }
+
+  inline T* operator->() { return get(); }
+  inline const T* operator->() const { return get(); }
+
+  inline Ptr<T> operator*() { return Ptr<T>(this); }
+
+private:
+  KJ_DISALLOW_COPY(Uniq);
+  
+  inline Uniq(T&& t): t(kj::mv(t)) {}
+
+  T t;
+
+#ifdef KJ_ASSERT_PIN_PTRS  
+  uint64_t ptrs = 0;
+#endif
+
+  inline T* get() { return &t; }
+  inline const T* get() const { return &t; }
+
+  template <typename X, typename... Params>
+  friend Uniq<X> uniq(Params&&... params);
+
+  template <typename X>
+  friend class Ptr;
+};
+
+template <typename T, typename... Params>
+inline Uniq<T> uniq(Params&&... params) {
+  return Uniq<T>(T(kj::fwd<Params>(params)...));
+}
+
+
+template <typename T>
+class Uniq<kj::Own<T>> {
+  // Uniq<kj::Own<T>> represent fully owned objects on the heap.
+
+public:
+  inline Uniq(kj::Own<T>&& t): t(kj::mv(t)) {}
+
+  inline Uniq(Uniq<kj::Own<T>>&& other): t(kj::mv(other.t)) {
+#ifdef KJ_ASSERT_PIN_PTRS  
+    KJ_IREQUIRE(other.ptrs == 0, "Uniq with active references moved");
+#endif
+  }
+
+  inline ~Uniq() {
+#ifdef KJ_ASSERT_PIN_PTRS  
+    KJ_IREQUIRE(ptrs == 0, "Uniq with active references destroyed");
+#endif
+  }
+
+  inline T* operator->() { return get(); }
+  inline const T* operator->() const { return get(); }
+
+  inline Ptr<T> operator*() { return Ptr<T>(this); }
+
+private:
+  KJ_DISALLOW_COPY(Uniq);
+  
+  kj::Own<T> t;
+#ifdef KJ_ASSERT_PIN_PTRS  
+  uint64_t ptrs = 0;
+#endif
+
+  inline T* get() { return t.get(); }
+  inline const T* get() const { return t.get(); }
+
+  template <typename X>
+  friend class Ptr;
+};
+
+// =======================================================================================
+// Ptr<T>
+
+template <typename T>
+class Ptr {
+public:
+  inline ~Ptr() {
+#ifdef KJ_ASSERT_PIN_PTRS  
+    (*ptrs)--;
+#endif
+  }
+
+  inline T* operator->() { return get(); }
```

Comment:
I think we need a way, given a `Ptr<T>`, to get a `Ptr` pointing at one of `T`'s members. It should share the same count as the parent, of course.

I am not sure what the syntax should look like. It could be a template where one of the template args is a pointer-to-member, but that'd be pretty verbose...